tools/autograd/derivatives.yaml

```diff
@@ -1727,6 +1727,15 @@
   self: leaky_relu_backward(grad, result, negative_slope, true)
   result: auto_element_wise
 
+- name: logical_not(Tensor self) -> Tensor
+  output_differentiability: [False]
+
+- name: logical_xor(Tensor self, Tensor other) -> Tensor
+  output_differentiability: [False]
+
+- name: logical_and(Tensor self, Tensor other) -> Tensor
+  output_differentiability: [False]
```

Comment:
@soulitzer what's the best way to declare these ops as non-differentiable? I guess the best case scenario would be if we can error out immediately in the forward pass, if someone tries to call `torch.logical_and(torch.ones(2, requres_grad=True))`.

I'm getting codegen errors that say `RuntimeError: output_differentiability=False for inplace operation (version_counter won't get updated)`. I see another option from the docs is to add these ops to the `DONT_REQUIRE_DERIVATIVE` in `gen_variable_type`?

Reply:
It is pretty counterintuitive but declaring these ops as non-differentiable actually makes it so that they do _not_ error when someone passes in a tensor that has `requires_grad=True`. The default behavior when no formula is specified is actually to error immediately in the forward pass.

A good example of why that is the case is `eq`, we wouldn't want users to have to `detach` everytime they want to compare two tensors. Maybe it is less clear what we really want to do in the `logical_{and,or}` cases, but adding it would be good to avoid bc-breaking issues.

I'm actually wondering how outdated that error is, since we update the version counter in the ADInplaceOrView kernel now. Perhaps we should just remove it?

It is also weird that `eq` does a similar thing (i.e., it has a in-place variant), but doesn't run into the error. I'm not sure why `eq` is able to work, but if you look into that perhaps you can emulate how it is done there.

Adding to `DONT_REQUIRE_DERIVATIVE` should also work, so I guess we can do that as a easy workaround for now.

Reply:
> A good example of why that is the case is eq, we wouldn't want users to have to detach everytime they want to compare two tensors.

Ah okay, then keeping the current behavior of not immediately erroring out in the forward pass sounds reasonable, to not be BC-breaking.

> Adding to DONT_REQUIRE_DERIVATIVE should also work, so I guess we can do that as a easy workaround for now.

I'll try this for now, thanks! I looked at `eq` and I don't think I'm doing anything differently for `logical_and/xor/not`.